File: aoc15/ocaml/day_20/day_20.ml
let input = int_of_string (Input.read "input") in
let n = input / 2 in

let deliver presents limit =
  let houses = Array.make (n + 1) 0 in
  for elf = 1 to n do
    let h = ref elf in
    let i = ref 0 in
    while !h < n && (limit = None || !i < Option.get limit) do
      houses.(!h) <- houses.(!h) + (elf * presents);
      h := !h + elf;
      i := !i + 1
    done
  done;
  houses
in

let lowest houses =
  let house, _ =
    Option.get
      (Array.find_opt
         (fun (_, v) -> v > input)
         (Array.mapi (fun h v -> (h, v)) houses))
  in
  house
in
let first = lowest (deliver 10 None) in
let second = lowest (deliver 11 (Some 50)) in
Printf.printf "First: %d\n" first;
Printf.printf "Second: %d\n" second

File: aoc15/go/day_20/day_20.go
package main

import (
	"fmt"
	"os"
	"slices"
	"strconv"
	"strings"
)

func deliver(houses []int, elf, presents, limit int) {
	var upper int
	if limit != 0 {
		upper = min(elf*limit+1, len(houses))
	} else {
		upper = len(houses)
	}
	for e := elf; e < upper; e += elf {
		houses[e] += elf * presents
	}
}

func run(target, presents, limit int) int {
	houses := make([]int, target/2)
	for elf := 1; elf < target/2; elf++ {
		deliver(houses, elf, presents, limit)
	}
	return slices.IndexFunc(houses, func(h int) bool { return h >= target })
}

func main() {
	input, _ := os.ReadFile("input")
	target, _ := strconv.Atoi(strings.TrimSpace(string(input)))

	fmt.Println("First:", run(target, 10, 0))
	fmt.Println("Second:", run(target, 11, 50))
}
